### PR TITLE
Fix java_stub_template.txt for cygwin environment

### LINKED ISSUE #20065: Java binary startup script broken when targeting windows (cygwin)
### Description of the bug:

I am building Java binaries on Linux which target Windows. The workflow is essentially build an archive containing the runfiles for my binary, copy that onto windows, extract and then run via the standard startup script which looks like it is designed to support cygwin. However the script is broken in a number of ways when I try to run on cygwin.

### Which category does this issue belong to?

Java Rules

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I've created a minimal repository for reproducing the issue:

1. Clone https://github.com/th0masb/bazel-java-startup-bug on a Linux environment
2. Run `bazel build --platforms //:windows_x86_64 :dist.tar`
3. Copy the resulting dist.tar archive to a windows machine with cygwin
4. Unpack the dist.tar in cygwin
5. Execute the entrypoint script from cygwin, it should fail. Execute the script with `--classpath_limit=1`, it should fail with a different reason.

Sample output:
```
$ ./.runfiles/__main__/main
Error: Could not find or load main class Main
Caused by: java.lang.ClassNotFoundException: Main

$ ./.runfiles/__main__/main --classpath_limit=1
./.runfiles/__main__/main: line 366: D:\Tools\Tom\.runfiles/remotejdk17_win/bin/java/jar.exe: No such file or directory
./.runfiles/__main__/main: ERROR: /cygdrive/d/Tools/Tom/./.runfiles/__main__/main failed because D:\Tools\Tom\.runfiles/remotejdk17_win/bin/java/jar.exe failed
```

### Which operating system are you running Bazel on?

Host + Execution Linux (Pop!_OS 22.04 LTS x86_64), target is Windows x86_64

### What is the output of `bazel info release`?

release 6.4.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

_No response_

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

Unsure

### Have you found anything relevant by searching the web?

No

### Any other information, logs, or outputs that you want to share?

I have identified how this could be fixed, the changes should be fairly minimal and low risk. I am happy to try and open a PR with a fix myself.

### PR DESCRIPTION
**Pull request by th0masb**
closes #20065 

Should be low risk change, and fix cross building java binary targets to run on Windows using cygwin.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt [hunk 1]
```diff
@@ -241,6 +241,13 @@ fi
 # Set JAVABIN to the path to the JVM launcher.
 %javabin%
 
+if is_windows; then
+  # Cygwin will not accept any backslashes in JAVABIN when trying to execute it.
+  JAVABIN="$(cygpath --unix $JAVABIN)"
+  # Ensure we have .exe suffix, JARBIN resolution fails without it.
```
Comment (fmeum) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt:247]:
```suggestion
  # Ensure we have .exe suffix, JAVABIN resolution fails without it.
```

Do you know why this is needed? The path to the java binary should be set correctly by the toolchain with https://github.com/bazelbuild/rules_java/pull/133. Do you still need this file with Bazel built from master?

File: src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt [hunk 2]
```diff
@@ -255,8 +262,14 @@ if [[ "$SINGLEJAR" == 1 ]]; then
...
   # Create the shortest classpath we can, by making it relative if possible.
   RUNPATH="${JAVA_RUNFILES}/%workspace_prefix%"
-  RUNPATH="${RUNPATH#$PWD/}"
+  # Keep absolute paths on windows, so we can correctly support both ':' and ';' as separators
+  if ! is_windows; then
+    RUNPATH="${RUNPATH#$PWD/}";
+  fi
   CLASSPATH=%classpath%
+  if is_windows; then
+    CLASSPATH="$(sed -E 's/:([A-Z]):/;\1:/g' <<< "$CLASSPATH")"
```
Comment (fmeum) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt:271]: Could you add a comment explaining what this does?

File: src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt (hunk 1 above)
Comment (th0masb) [reply src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt:247]: Yeah I did mean`JARBIN`, not `JAVABIN`.  You can see here [java_stub_template.txt#L361-L364](https://github.com/bazelbuild/bazel/blob/8c70d7b4ad5a9299f8e0cf7d2209fca0a29fd96f/src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt#L361-L364) that on windows `JARBIN` resolution requires the .exe suffix.

File: src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt (hunk 2 above)
Comment (th0masb) [reply src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt:275]: Done

**Comment by fmeum** [comment]
@cushon Could you take a look at this PR?

**Comment by cushon** [comment]
> @cushon Could you take a look at this PR?

I'll defer to @hvadehra on this one.

**Comment by hvadehra** [comment]
See https://github.com/bazelbuild/bazel/issues/20065#issuecomment-1880801326 for why I don't think this the right direction to fix this. 

Closing this for now, but happy to continue the discussion on the issue and we can re-open this if required.

**Pull request closed without merging**